darshan-runtime/lib/darshan.h

```diff
@@ -387,7 +387,7 @@ static inline int darshan_core_disabled_instrumentation(void)
 }
 
 /* retrieve absolute wtime */
-static inline double darshan_core_wtime_absolute(void)
+static inline double darshan_core_wtime_absolute(struct timespec *tspec)
```

Comment:
Can you clarify why you needed to make the change to pass in timespec values to these functions? Not clear from reviewing just the changes made to Darshan's core, I haven't started reviewing the changes to specific instrumentation modules yet.

Reply:
Looking further into the POSIX module, it looks like the timespecs are expected by the LDMS Darshan connector APIs. I feel like there might be a simpler way to modify all of the timer logic, but I'm not exactly sure what LDMS is expecting in terms of timestamps.

Reply:
LDMS is collecting timeseries data of each individual operation for each module so it is expecting the absolute time in which they occur instead of the relative time (i.e. epoch instead of seconds). 

From what I've found in the Darshan code, it looks like all of the timer logic (i.e. the timespec struct) is within this darshan_core_wtime_absolute() function which returns a floating point value. In order to collect the absolute timestamp without disrupting Darshans timer logic, we figured the best way to do this was to pass the timespec values through the functions where LDMS is being called/used (i.e. in the macro functions). 

We can add a separate timespec struct inside one of the Darshan LDMS functions but this would affect Darshan's overhead and performance.

Reply:
I see, thanks for the explanation.

I'm sure you noticed, but Darshan does capture time in terms of seconds/nanoseconds (using a `struct timespec`), it's just that it converts it to floating point time relative to the application start. when doing measurements. I'll think about it some more, but we might have a couple of options for simplifying this a bit:

1. Just have Darshan timing functions return timespecs and appropriately convert those elsewhere where we just want floating point elapsed time.
2. Darshan captures and stores a `struct timespec` for the start time of the job, of which the relative timers should be offset from. Maybe we could expose that value and let some logic handle the conversion when sending to LDMS (i.e., convert given relative timer to an absolute timer using the job start time).

Reply:
I agree that there's probably a better way to sort this out, we were just trying to get LDMS implemented and working on Darshan without modifying or requesting significant changes to the original code. 

1. This seems to be the most straight forward but could be a bit time consuming to do because, from what I recall, there are multiple functions/macros in all modules that use the floating point elapsed time.
2.  This is a bit more complex but once working, it doesn't sound like a lot of code refactoring is required.

Right now, I can't really can't think of any better ideas. But I am fine with which ever option is the best for Darshan and don't mind making the necessary changes in this branch.

Reply:
@shanedsnyder I spoke with one of our LDMS team members who's very experienced in the coding and development of LDMS and, based on their experience, they suggest only having timespec values returned from all timing functions would be best (i.e. option 1). LDMS already has this code structure and, as far as I know, it's been fairly quick and straightforward to implement any new timing functions, conversions, units, etc. (i.e. there hasn't been any need for major code changes).

I don't know if there were any new ideas about what to do with ```struct timespec``` or the future plans for Darshan but thought I'd let you know what we were thinking.

Reply:
I'm super duper late to this conversation, but we were chatting today about how to wrap up this PR now that Darshan 3.4.3 is finally out the door and I was curious about the change set :)

I have a strong preference not to alter the darshan wtime functions.  That change (to effectively make the function return timestamps in two formats) is making the diff a lot bigger than it ought to be.  For better or worse, relative floating point time stamps are deeply ingrained in Darshan at this point.

I would actually advocate for option 2).  The darshan.h header already has a global offset value (`__darshan_core_wtime_offset`) available in double format (so we can get back to an absolute value while still in floating point format without worrying about overflow) and it doesn't take many floating point operations from there to get back to a timespec.  Someone needs to sanity check this because I try to run it, but I think something like this might work in darshan.h:

```
static inline struct timespec darshan_core_abs_timespec_from_wtime(double t)
{
    struct timespec tp;

    /* add time offset back to get to absolute time */
    t += __darshan_core_time_offset;

    /* cast to integer type, deliberately truncating after decimal point to get whole number seconds */
    tp.tv_sec = (time_t)t;
    /* subtract seconds out, then multiply remainder to nsecs and cast */
    tp.tv_nsec = (time_t)((t - (double)tp.tv_sec) * 1.0e-9);
    return(tp);
}
```

If we had this, then the new ldms_send function could just use the existing available darshan wtime values from an API point of view and then call the above internally to translate to timespec before handing it to LDMS.  The code would look cleaner and I think we would have a difficult time measuring the performance impact of the conversion.

Reply:
Thanks, @carns for the code snippet. I think what you propose here makes a lot of sense. Let us know if this makes sense and/or if you could use some help @Snell1224

Reply:
@carns @shanedsnyder Thank you for the input and follow up. Yes, this makes sense to me and after reviewing the darshan code again, I agree that it would be much cleaner and easier to add this function than changing any of the existing darshan functions. I will go ahead and implement this code snippet, test it and update this PR with the changes. I'll reach out if I come across any issues or need any help. Thanks!